torch_geometric/nn/conv/updated_gcn.py

```diff
@@ -0,0 +1,253 @@
+                        edge_index, edge_weight, x.size(self.node_dim),
+                        self.improved, self.add_self_loops)
+                    if self.cached:
+                        self._cached_edge_index = (edge_index, edge_weight)
+                else:
+                    edge_index, edge_weight = cache[0], cache[1]
+
+            elif isinstance(edge_index, SparseTensor):
+                cache = self._cached_adj_t
+                if cache is None:
+                    edge_index = gcn_norm(  # yapf: disable
+                        edge_index, edge_weight, x.size(self.node_dim),
+                        self.improved, self.add_self_loops)
+                    if self.cached:
+                        self._cached_adj_t = edge_index
+                else:
+                    edge_index = cache
+
+        if not self.explain:
+            x = self.lin(x)
+
+        # propagate_type: (x: Tensor, edge_weight: OptTensor)
+        out = self.propagate(edge_index, x=x, edge_weight=edge_weight,
+                             message_scale=message_scale,
+                             message_replacement=message_replacement,
+                             size=None)
+
+        if self.bias is not None and not False and not self.explain:
+            out += self.bias
+
+        return out
+
+    def message(self, x_i: Tensor, x_j: Tensor, edge_index: Tensor,
+                edge_weight: OptTensor, message_scale,
+                message_replacement) -> Tensor:
+        if self.explain:
+            basis_messages = x_j if edge_weight is None else edge_weight.view(
+                -1, 1) * x_j
+            basis_messages = self.convert(basis_messages)
+            basis_messages = self.process_message(basis_messages)
+
+            if message_scale is not None:
+                basis_messages = basis_messages * message_scale.unsqueeze(-1)
+
+                if message_replacement is not None:
+                    if basis_messages.shape == message_replacement.shape:
+                        basis_messages = basis_messages + \
+                                         (1 - message_scale).unsqueeze(
+                                             -1) * message_replacement
+                    else:
+                        basis_messages = basis_messages + \
+                                         ((1 - message_scale).unsqueeze(-1) *
+                                          message_replacement.unsqueeze(0))
+
+            self.latest_messages = basis_messages
+            self.latest_source_embeddings = x_j
+            self.latest_target_embeddings = x_i
+
+            return basis_messages
+
+        return x_j if edge_weight is None else edge_weight.view(-1, 1) * x_j
+
+    # functions (get_latest_source_embeddings, get_latest_target_embeddings,
+    # and get_latest_messages) are necessary to compute edge_weights
+    # through hard_concrete_distribution
+    def get_latest_source_embeddings(self):
```

Comment:
I think it is save to drop these functions and just access, *e.g.*, `self.latest_source_embeddings`.

Reply:
As mentioned in the above comments that these three `get_latest_messages()`, `get_latest_source_embeddings()`, and `get_latest_target_embeddings()` functions act as containers for `messages`, `source_embeddings`, and `target_embeddings` respectively for each layer in the trained GNN Model. Since these functions help saving these (message, embeddings for both source and target nodes for each edge) for each layer, is it safe to drop these functions and just directly access `self.latest_messages`, `self.latest_source_embeddings`, and `self.latest_target_embeddings`?? What do you suggest??

Reply:
I don't see where these functions are used to save these features. I think it is safe to drop them and just access the features inside `GraphMask` directly.

Reply:
Agreed!!! I am also considering to drop these functions (`get_latest_messages()`, `get_latest_source_embeddings()`, and `get_latest_target_embeddings()`) and directly access these three instance attributes (`self.latest_messages`, `self.latest_source_embeddings`, and `self.latest_target_embeddings`) inside `GraphMask`.